src/SwcCompiler.ts

```diff
@@ -0,0 +1,235 @@
+
+  constructor() {
+    this.groups = new Map();
+  }
+
+  removeFile(filename: string) {
+    for (const [root, files] of this.groups.entries()) {
+      if (files.get(filename)) {
+        files.delete(filename);
+      }
+    }
+  }
+
+  addFile(file: CompiledFile) {
+    let group = this.groups.get(file.root);
+    if (!group) {
+      group = new Map();
+      this.groups.set(file.root, group);
+    }
+    group.set(file.filename, file);
+  }
+
+  group(filename: string): Group | undefined {
+    for (const [root, files] of this.groups.entries()) {
+      if (files.get(filename)) {
+        return { root, files: Array.from(files.values()) };
+      }
+    }
+  }
+}
+
+/** Implements TypeScript building using swc */
+export class SwcCompiler implements Compiler {
+  private compiledFiles: CompiledFiles;
+  private invalidatedFiles: Set<string>;
+  constructor(readonly workspaceRoot: string, readonly outDir: string) {
+    this.compiledFiles = new CompiledFiles();
+    this.invalidatedFiles = new Set();
+  }
+
+  async invalidateBuildSet() {
+    this.compiledFiles = new CompiledFiles();
+  }
+
+  async compile(filename: string): Promise<void> {
+    const existingGroup = this.compiledFiles.group(filename);
+
+    if (existingGroup) {
+      await this.buildFile(filename, existingGroup.root);
+    } else {
+      await this.buildGroup(filename);
+    }
+
+    return;
+  }
+
+  async fileGroup(filename: string) {
+    const contents: Record<string, string> = {};
+    const group = this.compiledFiles.group(filename);
+
+    if (!group) {
+      throw new Error(await this.missingDestination(filename));
+    }
+
+    for (const file of group.files) {
+      contents[file.filename] = file.destination;
+    }
+
+    return contents;
+  }
+
+  private async getModule(filename: string) {
+    const root = findRoot(filename);
+    const config = await projectConfig(root);
+    const globs = [...this.fileGlobPatterns(config), ...this.ignoreFileGlobPatterns(config)];
+
+    log.debug("searching for filenames", { config, root, globs });
+
+    let fileNames = await globby(globs, { cwd: root, absolute: true });
+
+    if (process.platform === "win32") {
+      fileNames = fileNames.map((fileName) => fileName.replace(/\//g, "\\"));
+    }
+
+    return { root, fileNames, config };
+  }
+
+  private async buildFile(filename: string, root: string): Promise<CompiledFile> {
+    const file = await transformFile(filename, {
+      cwd: root,
+      filename: filename,
+      root: this.workspaceRoot,
+      rootMode: "root",
+      sourceMaps: "inline",
+      inlineSourcesContent: true,
+      env: {
+        targets: {
+          node: 16,
+        },
+      },
+      jsc: {
+        parser: {
+          syntax: "typescript",
+          decorators: true,
+          dynamicImport: true,
+        },
+        target: "es2020",
+      },
+      module: {
+        type: "commonjs",
+        lazy: true,
+      },
+    }).then(
```

Comment:
Don't think you should have to use a `then` here, you can just put the contents of the function below on the next lines. Passing an `async` function to a `.then` also doesn't have any special properties and creates what's called a floating promise which is a no no, it is subtle and annoying. If you're using `async`/`await` in general there should be no need to create `new Promise`es explicitly or use `.then`s!

Reply:
Oh good to know, thanks!